ipykernel/subshell_manager.py

```diff
@@ -0,0 +1,236 @@
+
+        self._poller = zmq.Poller()
+        self._poller.register(self._parent_shell_channel_socket, zmq.POLLIN)
+
+    def close(self):
+        """Stop all subshells and close all resources."""
+        assert current_thread().name == "Shell channel"
+
+        for socket in (
+            self._control_shell_channel_socket,
+            self._control_other_socket,
+            self._parent_shell_channel_socket,
+            self._parent_other_socket,
+        ):
+            if socket is not None:
+                socket.close()
+
+        with self._lock:
+            while True:
+                try:
+                    _, subshell = self._cache.popitem()
+                except KeyError:
+                    break
+                self._stop_subshell(subshell)
+
+    def get_control_other_socket(self):
+        return self._control_other_socket
+
+    def get_other_socket(self, subshell_id: str | None):
+        """Return the other inproc pair socket for a subshell.
+
+        This socket is accessed from the subshell thread.
+        """
+        if subshell_id is None:
+            return self._parent_other_socket
+        with self._lock:
+            return self._cache[subshell_id].thread._pair_socket
+
+    def get_shell_channel_socket(self, subshell_id: str | None):
+        """Return the shell channel inproc pair socket for a subshell.
+
+        This socket is accessed from the shell channel thread.
+        """
+        if subshell_id is None:
+            return self._parent_shell_channel_socket
+        with self._lock:
+            return self._cache[subshell_id].shell_channel_socket
+
+    def list_subshell(self) -> list[str]:
+        """Return list of current subshell ids.
+
+        Can be called by any subshell using %subshell magic.
+        """
+        with self._lock:
+            return list(self._cache)
+
+    async def listen_from_control(self, subshell_task):
+        """Listen for messages on the control inproc socket, handle those messages and
+        return replies on the same socket.  Runs in the shell channel thread.
+        """
+        assert current_thread().name == "Shell channel"
+
+        socket = self._control_shell_channel_socket
+        while True:
+            request = await socket.recv_json()
+            reply = await self._process_control_request(request, subshell_task)
+            await socket.send_json(reply)
+
+    async def listen_from_subshells(self):
+        """Listen for reply messages on inproc sockets of all subshells and resend
+        those messages to the client via the shell_socket.
+
+        Runs in the shell channel thread.
+        """
+        assert current_thread().name == "Shell channel"
+
+        while True:
+            for socket, _ in self._poller.poll(0):
+                msg = await socket.recv_multipart(copy=False)
+                self._shell_socket.send_multipart(msg)
+
+            # Yield to other tasks.
+            await sleep(0)
```

Comment:
This loop leads to 100% CPU usage, we need to find another way.

Reply:
Interesting. On my macOS dev machine this is fine, but I've confirmed that it is a problem on Linux.

Reply:
I've changed the implementation in 97e3e91e2a. I've removed the `sleep(0)` and `poll(0)` calls and now use an async zmq `Poller` in an `anyio` task. When a subshell is created or deleted this task is cancelled using an `anyio.Event` are rescheduled with the updated list of subshells. With this I no longer see 100% CPU load on Linux or macOS.

Reply:
I've changed the implementation of this piece of code again. My use of `await poller.poll()` turned out to cause problems on python < 3.10 and on Windows. Now I am avoiding use of `zmq.Poller` and instead using a separate anyio task for each subshell to listen for reply messages and send them out to the client via the shell channel. This seems to be more robust on older python and Windows. I've also replaced the use of an `anyio.Event` with a memory object stream instead (essentially an anyio async queue).